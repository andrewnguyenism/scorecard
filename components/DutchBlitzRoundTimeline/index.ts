export { DutchBlitzRoundTimeline } from './DutchBlitzRoundTimeline'

export type { DutchBlitzRound } from './DutchBlitzRoundTimeline'
