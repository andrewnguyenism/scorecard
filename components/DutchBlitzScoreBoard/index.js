export {
  DutchBlitzScoreBoardHeader,
  DutchBlitzScoreBoardPlayer,
  DutchBlitzScoreBoard,
} from "./DutchBlitzScoreBoard";
