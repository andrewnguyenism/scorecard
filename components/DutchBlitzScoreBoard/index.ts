export { DutchBlitzScoreBoard } from './DutchBlitzScoreBoard'
export { DutchBlitzScoreBoardHeader } from './DutchBlitzScoreBoardHeader'
export { DutchBlitzScoreBoardPlayer } from './DutchBlitzScoreBoardPlayer'
