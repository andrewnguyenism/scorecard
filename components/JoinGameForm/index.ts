export { JoinGameFormError, JoinGameForm } from './JoinGameForm'
