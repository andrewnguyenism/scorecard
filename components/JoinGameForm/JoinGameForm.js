import { useEffect, useState } from "react";

import { useRouter } from "next/router";

import { useUser } from "../../context/UserContext";
import firebase from "../../firebase/client";

import { GameIdWidget } from "../GameIdWidget";

export const JoinGameForm = ({ gameId, game, playerCount }) => {
  const { user } = useUser();
  const router = useRouter();

  const [status, setStatus] = useState("idle");
  const [gameCode, setGameCode] = useState("");
  const [gameInfo, setGameInfo] = useState(null);
  const [noGameError, setNoGameError] = useState(false);
  const [joinGameName, setJoinGameName] = useState("");
  const [duplicateNameError, setDuplicateNameError] = useState(false);

  useEffect(() => {
    const checkGameExists = async () => {
      if (gameCode.length === 5) {
        const snapshot = await firebase.database().ref(`games/${gameCode}`).once('value');
        if (snapshot.val()) {
          setGameInfo(snapshot.val());
        } else {
          setNoGameError(true);
        }
      } else {
        setGameInfo(null);
        if (noGameError) {
          setNoGameError(false);
        }
      }
    }
    checkGameExists();
  }, [gameCode]);

  const joinGame = async () => {
    setStatus("joining");
    const gameStatus = await firebase
      .database()
      .ref(`games/${gameCode}/gameStatus`)
      .once("value");
      
    if (!gameId && !gameStatus.val()) {
      setNoGameError(true);
      setStatus("error");
      return false;
    }

    const players = await firebase
      .database()
      .ref(`games/${gameId || gameCode}/players`)
      .once("value");
    if (
      Object.values(players.val())
        .find((player) => player.toLowerCase() === joinGameName.toLowerCase())
    ) {
      setDuplicateNameError(true);
    } else {
      const updates = {
        [`/games/${gameId || gameCode}/players/${user.uid}`]: joinGameName,
        [`/players/${gameId || gameCode}/${user.uid}`]: {
          name: joinGameName,
          dutchScore: 0,
          blitzScore: 0,
          totalScore: 0,
          createdAt: firebase.database.ServerValue.TIMESTAMP,
          updatedAt: firebase.database.ServerValue.TIMESTAMP,
        },
        [`/users/${user.uid}/${gameId || gameCode}`]: {
          name: joinGameName,
          createdAt: firebase.database.ServerValue.TIMESTAMP,
          updatedAt: firebase.database.ServerValue.TIMESTAMP,
        },
      };
      await firebase.database().ref().update(updates);
      setStatus("joined");
      router.push(`/game/${gameId || gameCode}`);
    }
  };

  return (
    <div className="rounded-2xl border-2 overflow-hidden m-4 p-6 text-center">
      {(game || gameInfo?.game) && (
        <div className="font-bold uppercase mb-2">
          {(game || gameInfo?.game) === "dutch-blitz" && "Dutch Blitz"}
        </div>
      )}
      {gameId && (
        <div className="mb-4 text-lg">
          <GameIdWidget gameId={gameId} />
        </div>
      )}
      {(playerCount || (gameInfo && Object.keys(gameInfo.players).length > 0)) && (
        <div className="mb-4 text-sm">
          {playerCount || Object.keys(gameInfo.players).length} player
          {`${(playerCount || Object.keys(gameInfo.players).length) > 1 ? "s" : ""}`} in
          game
        </div>
      )}
      {!gameId && (
        <div className="flex flex-col mb-4">
          <label
            className="uppercase font-bold text-gray-900 mb-2"
            htmlFor="code"
          >
            Game Code
          </label>
          <input
            className={`border py-2 px-3 text-grey-darkest uppercase text-xl text-center font-mono tracking-widest ${
              noGameError && "border-red-500"
            }`}
            id="code"
            maxLength="5"
            type="text"
            placeholder="AB123"
            onChange={(event) => {
              setGameCode(event.currentTarget.value.toUpperCase());
              if (noGameError) {
                setNoGameError(false);
              }
            }}
            value={gameCode}
          />
          {noGameError && (
            <div className="text-sm mt-1">
              Game doesn't exist, make sure you got the code right.
            </div>
          )}
        </div>
      )}
      <div className="flex flex-col mb-4">
        <label
          className="uppercase font-bold text-gray-900 mb-2"
          htmlFor="name"
        >
          Display Name
        </label>
        <input
          className={`border py-2 px-3 text-grey-darkest text-center ${
            duplicateNameError && "border-red-500"
          }`}
          disabled={gameInfo?.players[user.uid]}
          id="name"
          type="text"
          placeholder={gameInfo?.players[user.uid] || "Your Name"}
          onChange={(event) => {
            setJoinGameName(event.currentTarget.value);
            if (duplicateNameError) {
              setDuplicateNameError(false);
            }
          }}
          value={joinGameName}
        />
        {duplicateNameError && (
          <div className="text-sm mt-1">
            Name already used, try another one.
          </div>
        )}
      </div>
      {gameInfo?.players[user.uid] ? (
        <div className="text-sm mt-1">
          You're already in this game!
        </div>
      ) : (
        <button
          className="block bg-indigo-800 hover:bg-indigo-700 text-white uppercase text-sm font-semibold mx-auto px-4 py-2 rounded-2xl"
          onClick={() => joinGame()}
        >
          Join
        </button>
      )}
    </div>
  );
};
