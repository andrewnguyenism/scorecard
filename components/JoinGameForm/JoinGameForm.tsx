import { FunctionComponent, useEffect, useState } from "react";

import { useRouter } from "next/router";

import firebase from "../../firebase/client";
import { GameIdWidget } from "../GameIdWidget";
import { DutchBlitzPlayer } from "../DutchBlitzScoreBoard";

export interface GameInfo {
  createdAt: number;
  currentRound: string;
  game: string;
  gameStatus: "playing" | "finished";
  owner: string;
  players: {
    [id: string]: string;
  };
  roundStatus: string;
  rounds: number;
  updatedAt: number;
}

interface Props {
  currentUserId: string;
  gameId?: string;
  game?: string;
  playerCount?: number;
}

export const JoinGameForm: FunctionComponent<Props> = ({
  currentUserId,
  gameId,
  game,
  playerCount,
}) => {
  const router = useRouter();

  const [status, setStatus] = useState("idle");
  const [gameCode, setGameCode] = useState("");
  const [gameInfo, setGameInfo] = useState<GameInfo | null>(null);
  const [noGameError, setNoGameError] = useState(false);
  const [joinGameName, setJoinGameName] = useState("");
  const [duplicateNameError, setDuplicateNameError] = useState(false);

  useEffect(() => {
    const checkGameExists = async () => {
      if (gameCode.length === 5) {
        const snapshot = await firebase
          .database()
          .ref(`games/${gameCode}`)
          .once("value");
        if (snapshot.val()) {
          setGameInfo(snapshot.val());
        } else {
          setNoGameError(true);
        }
      } else {
        setGameInfo(null);
        if (noGameError) {
          setNoGameError(false);
        }
      }
    };
    checkGameExists();
  }, [gameCode, noGameError]);

  const joinGame = async () => {
    setStatus("joining");
    const gameStatus = await firebase
      .database()
      .ref(`games/${gameCode}/gameStatus`)
      .once("value");

    if (!gameId && !gameStatus.val()) {
      setNoGameError(true);
      setStatus("error");
      return false;
    }

    const players = await firebase
      .database()
      .ref(`games/${gameId || gameCode}/players`)
      .once("value");
    if (
      Object.values(players.val() as DutchBlitzPlayer).find(
        (player) => player.toLowerCase() === joinGameName.toLowerCase()
      )
    ) {
      setDuplicateNameError(true);
    } else {
      const updates = {
        [`/games/${gameId || gameCode}/players/${currentUserId}`]: joinGameName,
        [`/players/${gameId || gameCode}/${currentUserId}`]: {
          name: joinGameName,
          dutchScore: 0,
          blitzScore: 0,
          totalScore: 0,
          createdAt: firebase.database.ServerValue.TIMESTAMP,
          updatedAt: firebase.database.ServerValue.TIMESTAMP,
        },
        [`/users/${currentUserId}/${gameId || gameCode}`]: {
          name: joinGameName,
          createdAt: firebase.database.ServerValue.TIMESTAMP,
          updatedAt: firebase.database.ServerValue.TIMESTAMP,
        },
      };
      await firebase.database().ref().update(updates);
      setStatus("joined");
      router.push(`/game/${gameId || gameCode}`);
    }
  };

  const numberOfPlayers =
    playerCount || (gameInfo && Object.keys(gameInfo.players).length);

  return (
    <div className="rounded-2xl border-2 overflow-hidden m-4 p-6 text-center">
      {(game || gameInfo?.game) && (
        <div className="font-bold uppercase mb-2">
          {(game || gameInfo?.game) === "dutch-blitz" && "Dutch Blitz"}
        </div>
      )}
      {gameId && (
        <div className="mb-4 text-lg">
          <GameIdWidget gameId={gameId} />
        </div>
      )}
      {numberOfPlayers && (
        <div className="mb-4 text-sm">
          {numberOfPlayers} player
          {`${numberOfPlayers > 1 ? "s" : ""}`} in game
        </div>
      )}
      {!gameId && (
        <div className="flex flex-col mb-4">
          <label
            className="uppercase font-bold text-gray-900 mb-2"
            htmlFor="code"
          >
            Game Code
          </label>
          <input
            className={`border py-2 px-3 text-grey-darkest uppercase text-xl text-center font-mono tracking-widest ${
              noGameError && "border-red-500"
            }`}
            id="code"
            maxLength={5}
            type="text"
            placeholder="AB123"
            onChange={(event) => {
              setGameCode(event.currentTarget.value.toUpperCase());
              if (noGameError) {
                setNoGameError(false);
              }
            }}
            value={gameCode}
          />
          {noGameError && (
            <div className="text-sm mt-1">
              Game doesn&apos;t exist, make sure you got the code right.
            </div>
          )}
        </div>
      )}
      <div className="flex flex-col mb-4">
        <label
          className="uppercase font-bold text-gray-900 mb-2"
          htmlFor="name"
        >
          Display Name
        </label>
        <input
          className={`border py-2 px-3 text-grey-darkest text-center ${
            duplicateNameError && "border-red-500"
          }`}
          disabled={!!gameInfo?.players[currentUserId]}
          id="name"
          type="text"
          placeholder={gameInfo?.players[currentUserId] || "Your Name"}
          onChange={(event) => {
            setJoinGameName(event.currentTarget.value);
            if (duplicateNameError) {
              setDuplicateNameError(false);
            }
          }}
          value={joinGameName}
        />
        {duplicateNameError && (
          <div className="text-sm mt-1">
            Name already used, try another one.
          </div>
        )}
      </div>
      {gameInfo?.players[currentUserId] ? (
        <div className="text-sm mt-1">You&apos;re already in this game!</div>
      ) : (
        <button
          className="block bg-indigo-800 hover:bg-indigo-700 text-white uppercase text-sm font-semibold mx-auto px-4 py-2 rounded-2xl"
          onClick={() => joinGame()}
        >
          Join
        </button>
      )}
    </div>
  );
};
