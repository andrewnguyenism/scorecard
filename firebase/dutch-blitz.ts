import firebase from './client'
import { getGameCurrentRound } from './game'
import { getUserPlayerInfo } from './player'
import { getUserRoundInfo } from './round'

export const callBlitz = async (
  gameId: string,
  userId: string
): Promise<boolean> => {
  const updates = {
    [`/games/${gameId}/roundStatus`]: 'blitz',
    [`/players/${gameId}/${userId}/blitz`]: true,
  }
  try {
    await firebase.database().ref().update(updates)
    return true
  } catch (err) {
    console.error(err)
    return false
  }
}

export const submitDutchBlitzScore = async (
  gameId: string,
  userId: string,
  dutchScore: number,
  blitzScore: number,
  lastSubmittedRoundId: string | null
): Promise<boolean> => {
  const currentRoundId = await getGameCurrentRound(gameId)
  if (currentRoundId) {
    const currentPlayerInfo = await getUserPlayerInfo(gameId, userId)
    let updates = {
      [`/rounds/${gameId}/${currentRoundId}/players/${userId}/dutchScore`]: dutchScore,
      [`/rounds/${gameId}/${currentRoundId}/players/${userId}/blitzScore`]: blitzScore,
      [`/rounds/${gameId}/${currentRoundId}/players/${userId}/totalScore`]:
        dutchScore - blitzScore * 2,
      [`/rounds/${gameId}/${currentRoundId}/players/${userId}/updatedAt`]: firebase
        .database.ServerValue.TIMESTAMP,
    }
    if (lastSubmittedRoundId !== currentRoundId) {
      updates = {
        ...updates,
        [`/players/${gameId}/${userId}/dutchScore`]:
          (currentPlayerInfo?.dutchScore || 0) + dutchScore,
        [`/players/${gameId}/${userId}/blitzScore`]:
          (currentPlayerInfo?.blitzScore || 0) + blitzScore,
        [`/players/${gameId}/${userId}/totalScore`]:
          (currentPlayerInfo?.totalScore || 0) + (dutchScore - blitzScore * 2),
        [`/players/${gameId}/${userId}/updatedAt`]: firebase.database.ServerValue
          .TIMESTAMP,
        [`/players/${gameId}/${userId}/lastRoundSubmitted`]: currentRoundId,
      }
    } else {
      const currentRoundInfo = await getUserRoundInfo(
        gameId,
        currentRoundId,
        userId
      )
      updates = {
        ...updates,
        [`/players/${gameId}/${userId}/dutchScore`]:
          (currentPlayerInfo?.dutchScore || 0) -
          (currentRoundInfo?.dutchScore || 0) +
          dutchScore,
        [`/players/${gameId}/${userId}/blitzScore`]:
          (currentPlayerInfo?.blitzScore || 0) -
          (currentRoundInfo?.blitzScore || 0) +
          blitzScore,
        [`/players/${gameId}/${userId}/totalScore`]:
          (currentPlayerInfo?.totalScore || 0) -
          (currentRoundInfo?.totalScore || 0) +
          (dutchScore - blitzScore * 2),
        [`/players/${gameId}/${userId}/updatedAt`]: firebase.database.ServerValue
          .TIMESTAMP,
      }
    }
    try {
      await firebase.database().ref().update(updates)
      return true
    } catch (err) {
      console.error(err)
      return false
    }
  } else {
    return false
  }
}
