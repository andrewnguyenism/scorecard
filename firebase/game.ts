import firebase from './client'

export enum ScorecardGame {
  DutchBlitz = 'dutch-blitz'
}

type GameStatus = 'finished' | 'playing'

interface GamePlayers {
  [id: string]: string
}

export interface GameInfo {
  createdAt: number
  currentRound: string
  game: ScorecardGame
  gameStatus: GameStatus
  owner: string
  players: GamePlayers
  roundStatus: string
  rounds: number
  updatedAt: number
}

interface CreateGameData {
  game: ScorecardGame
  name: string
  userId: string
}

interface JoinGameData {
  gameId: string
  name: string
  userId: string
}

export const createGame = async ({
  game,
  name,
  userId,
}: CreateGameData): Promise<string | null> => {
  const gameId = Math.random().toString(36).substr(2, 5).toUpperCase()
  const roundKey = firebase.database().ref(`rounds/${gameId}`).push().key
  const promises = [
    firebase
      .database()
      .ref(`games/${gameId}`)
      .set({
        currentRound: roundKey,
        game,
        gameStatus: 'playing',
        owner: userId,
        players: {
          [userId]: name,
        },
        rounds: 1,
        roundStatus: 'playing',
        voteStatus: null,
        createdAt: firebase.database.ServerValue.TIMESTAMP,
        updatedAt: firebase.database.ServerValue.TIMESTAMP,
      }),
    firebase.database().ref(`rounds/${gameId}/${roundKey}`).set({
      createdAt: firebase.database.ServerValue.TIMESTAMP,
      updatedAt: firebase.database.ServerValue.TIMESTAMP,
    }),
    firebase.database().ref(`players/${gameId}/${userId}`).set({
      name,
      dutchScore: 0,
      blitzScore: 0,
      totalScore: 0,
      createdAt: firebase.database.ServerValue.TIMESTAMP,
      updatedAt: firebase.database.ServerValue.TIMESTAMP,
    }),
    firebase.database().ref(`users/${userId}/${gameId}`).set({
      name,
      owner: true,
      createdAt: firebase.database.ServerValue.TIMESTAMP,
      updatedAt: firebase.database.ServerValue.TIMESTAMP,
    }),
  ]
  try {
    await Promise.all(promises)
    return gameId
  } catch (err) {
    console.error(err)
    return null
  }
}

export const joinGame = async ({
  gameId,
  name,
  userId,
}: JoinGameData): Promise<boolean> => {
  const updates = {
    [`/games/${gameId}/players/${userId}`]: name,
    [`/players/${gameId}/${userId}`]: {
      name: name,
      dutchScore: 0,
      blitzScore: 0,
      totalScore: 0,
      createdAt: firebase.database.ServerValue.TIMESTAMP,
      updatedAt: firebase.database.ServerValue.TIMESTAMP,
    },
    [`/users/${userId}/${gameId}`]: {
      name: name,
      createdAt: firebase.database.ServerValue.TIMESTAMP,
      updatedAt: firebase.database.ServerValue.TIMESTAMP,
    },
  }
  try {
    await firebase.database().ref().update(updates)
    return true
  } catch (err) {
    console.error(err)
    return false
  }
}

export const doesGameExist = async (gameId: string): Promise<boolean> => {
  try {
    const snapshot = await firebase
      .database()
      .ref(`games/${gameId}/createdAt`)
      .once('value')
    return snapshot.exists()
  } catch (err) {
    console.error(err)
    return false
  }
}

export const listenForGameInfoChange = (
  gameId: string,
  listener: (gameInfo: GameInfo | null) => void
): (() => void) => {
  const onChange = (snapshot: firebase.database.DataSnapshot) => {
    listener(snapshot.val())
  }
  firebase.database().ref(`games/${gameId}`).on('value', onChange)
  return () => firebase.database().ref(`games/${gameId}`).off('value', onChange)
}

export const getGameInfo = async (gameId: string): Promise<GameInfo | null> => {
  try {
    const snapshot = await firebase.database().ref(`games/${gameId}`).once('value')
    return snapshot.val()
  } catch (err) {
    console.error(err)
    return null
  }
}

export const getGameCurrentRound = async (gameId: string): Promise<string | null> => {
  try {
    const snapshot = await firebase
      .database()
      .ref(`games/${gameId}/currentRound`)
      .once('value')
    return snapshot.val()
  } catch (err) {
    console.error(err)
    return null
  }
}

export const getGameRounds = async (gameId: string): Promise<number | null> => {
  try {
    const snapshot = await firebase
      .database()
      .ref(`games/${gameId}/rounds`)
      .once('value')
    return snapshot.val()
  } catch (err) {
    console.error(err)
    return null
  }
}

export const getGamePlayers = async (gameId: string): Promise<GamePlayers | null> => {
  try {
    const snapshot = await firebase
      .database()
      .ref(`games/${gameId}/players`)
      .once('value')
    return snapshot.val()
  } catch (err) {
    console.error(err)
    return null
  }
}

export const setGameFinished = async (gameId: string): Promise<boolean> => {
  const updates = {
    [`/games/${gameId}/gameStatus`]: 'finished',
    [`/games/${gameId}/updatedAt`]: firebase.database.ServerValue.TIMESTAMP,
  }
  try {
    await firebase.database().ref().update(updates)
    return true
  } catch (err) {
    console.error(err)
    return false
  }
}

export const progressGameToNextRound = async (gameId: string): Promise<boolean> => {
  const newRoundKey = firebase.database().ref(`rounds/${gameId}`).push().key
  const updates = {
    [`/games/${gameId}/currentRound`]: newRoundKey,
    [`/games/${gameId}/roundStatus`]: 'playing',
    [`/games/${gameId}/rounds`]: (await getGameRounds(gameId) || 0) + 1,
    [`/rounds/${gameId}/${newRoundKey}`]: {
      createdAt: firebase.database.ServerValue.TIMESTAMP,
      updatedAt: firebase.database.ServerValue.TIMESTAMP,
    },
  }
  try {
    await firebase.database().ref().update(updates)
    return true
  } catch (err) {
    console.error(err)
    return false
  }
}