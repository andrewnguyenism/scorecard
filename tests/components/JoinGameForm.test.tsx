import user from '@testing-library/user-event'
import { render, screen } from '../test-utils'

import { JoinGameFormError, JoinGameForm } from '@/components/JoinGameForm'
import { ScorecardGame } from '@/firebase/game'

describe('JoinGameForm component', () => {
  it('renders when no gameId is passed', () => {
    const { asFragment } = render(
      <JoinGameForm />
    )
    expect(asFragment()).toMatchSnapshot()
  })

  it('renders when gameId, game, and playerCount is passed', () => {
    const { asFragment } = render(
      <JoinGameForm
        gameId="AB123"
        game={ScorecardGame.DutchBlitz}
        playerCount={4}
      />
    )
    expect(asFragment()).toMatchSnapshot()
  })

  it ('renders when duplicate name error is passed', async () => {
    const { asFragment } = render(
      <JoinGameForm error={JoinGameFormError.DuplicateName} />
    )

    await screen.findByText(/Name already used, try another one\./i)
    expect(asFragment()).toMatchSnapshot()
  })

  it ('renders when no game error is passed', async () => {
    const { asFragment } = render(
      <JoinGameForm error={JoinGameFormError.DuplicateName} />
    )

    await screen.findByText(/Name already used, try another one\./i)
    expect(asFragment()).toMatchSnapshot()
  })

  it('calls clearError after changing input', () => {
    const clearError = jest.fn()
    const gameId = 'AB123'

    render(
      <JoinGameForm error={JoinGameFormError.GameNotExists} clearError={clearError} />
    )

    user.type(screen.getByLabelText(/game code/i), gameId)

    expect(clearError).toBeCalled()
  })

  it('calls onGameCodeEntered after entering code', () => {
    const onGameCodeEntered = jest.fn()
    const gameId = 'AB123'

    render(
      <JoinGameForm onGameCodeEntered={onGameCodeEntered} />
    )

    user.type(screen.getByLabelText(/game code/i), gameId)

    expect(onGameCodeEntered).toBeCalledWith(gameId)
  })

  it('calls onJoinGameClick after clicking "Join"', () => {
    const onJoinGameClicked = jest.fn()
    const gameId = 'AB123'
    const name = 'Name'

    render(
      <JoinGameForm onJoinGameClicked={onJoinGameClicked} />
    )

    user.type(screen.getByLabelText(/game code/i), gameId)
    user.type(screen.getByLabelText(/name/i), name)
    user.click(screen.getByText(/join/i))

    expect(onJoinGameClicked).toBeCalledWith(name, gameId)
  })
})
