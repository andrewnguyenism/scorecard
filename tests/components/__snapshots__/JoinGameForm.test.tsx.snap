// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JoinGameForm component renders when duplicate name error is passed 1`] = `
<DocumentFragment>
  <div
    class="rounded-2xl border-2 overflow-hidden m-4 p-6 text-center"
  >
    <div
      class="flex flex-col mb-4"
    >
      <label
        class="uppercase font-bold text-gray-900 mb-2"
        for="code"
      >
        Game Code
      </label>
      <input
        class="border py-2 px-3 text-grey-darkest uppercase text-xl text-center font-mono tracking-widest false"
        id="code"
        maxlength="5"
        placeholder="AB123"
        type="text"
        value=""
      />
    </div>
    <div
      class="flex flex-col mb-4"
    >
      <label
        class="uppercase font-bold text-gray-900 mb-2"
        for="name"
      >
        Display Name
      </label>
      <input
        class="border py-2 px-3 text-grey-darkest text-center border-red-500"
        id="name"
        placeholder="Your Name"
        type="text"
        value=""
      />
      <div
        class="text-sm mt-1"
      >
        Name already used, try another one.
      </div>
    </div>
    <button
      class="block bg-indigo-800 hover:bg-indigo-700 text-white uppercase text-sm font-semibold mx-auto px-4 py-2 rounded-2xl"
    >
      Join
    </button>
  </div>
</DocumentFragment>
`;

exports[`JoinGameForm component renders when gameId, game, and playerCount is passed 1`] = `
<DocumentFragment>
  <div
    class="rounded-2xl border-2 overflow-hidden m-4 p-6 text-center"
  >
    <div
      class="font-bold uppercase mb-2"
    >
      Dutch Blitz
    </div>
    <div
      class="mb-4 text-lg"
    >
      <div
        class="cursor-pointer rounded-2xl bg-gray-300 hover:bg-gray-400 px-3 inline-flex items-center font-bold"
        role="button"
      >
        <span
          class="text-gray-800 text-xl font-mono font-bold tracking-widest"
        >
          AB123
        </span>
        <svg
          class="h-4 w-4 ml-0.5"
          fill="currentColor"
          viewBox="0 0 20 20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M8 2a1 1 0 000 2h2a1 1 0 100-2H8z"
          />
          <path
            d="M3 5a2 2 0 012-2 3 3 0 003 3h2a3 3 0 003-3 2 2 0 012 2v6h-4.586l1.293-1.293a1 1 0 00-1.414-1.414l-3 3a1 1 0 000 1.414l3 3a1 1 0 001.414-1.414L10.414 13H15v3a2 2 0 01-2 2H5a2 2 0 01-2-2V5zM15 11h2a1 1 0 110 2h-2v-2z"
          />
        </svg>
      </div>
    </div>
    <div
      class="mb-4 text-sm"
    >
      4 players in game
    </div>
    <div
      class="flex flex-col mb-4"
    >
      <label
        class="uppercase font-bold text-gray-900 mb-2"
        for="name"
      >
        Display Name
      </label>
      <input
        class="border py-2 px-3 text-grey-darkest text-center false"
        id="name"
        placeholder="Your Name"
        type="text"
        value=""
      />
    </div>
    <button
      class="block bg-indigo-800 hover:bg-indigo-700 text-white uppercase text-sm font-semibold mx-auto px-4 py-2 rounded-2xl"
    >
      Join
    </button>
  </div>
</DocumentFragment>
`;

exports[`JoinGameForm component renders when no game error is passed 1`] = `
<DocumentFragment>
  <div
    class="rounded-2xl border-2 overflow-hidden m-4 p-6 text-center"
  >
    <div
      class="flex flex-col mb-4"
    >
      <label
        class="uppercase font-bold text-gray-900 mb-2"
        for="code"
      >
        Game Code
      </label>
      <input
        class="border py-2 px-3 text-grey-darkest uppercase text-xl text-center font-mono tracking-widest false"
        id="code"
        maxlength="5"
        placeholder="AB123"
        type="text"
        value=""
      />
    </div>
    <div
      class="flex flex-col mb-4"
    >
      <label
        class="uppercase font-bold text-gray-900 mb-2"
        for="name"
      >
        Display Name
      </label>
      <input
        class="border py-2 px-3 text-grey-darkest text-center border-red-500"
        id="name"
        placeholder="Your Name"
        type="text"
        value=""
      />
      <div
        class="text-sm mt-1"
      >
        Name already used, try another one.
      </div>
    </div>
    <button
      class="block bg-indigo-800 hover:bg-indigo-700 text-white uppercase text-sm font-semibold mx-auto px-4 py-2 rounded-2xl"
    >
      Join
    </button>
  </div>
</DocumentFragment>
`;

exports[`JoinGameForm component renders when no gameId is passed 1`] = `
<DocumentFragment>
  <div
    class="rounded-2xl border-2 overflow-hidden m-4 p-6 text-center"
  >
    <div
      class="flex flex-col mb-4"
    >
      <label
        class="uppercase font-bold text-gray-900 mb-2"
        for="code"
      >
        Game Code
      </label>
      <input
        class="border py-2 px-3 text-grey-darkest uppercase text-xl text-center font-mono tracking-widest false"
        id="code"
        maxlength="5"
        placeholder="AB123"
        type="text"
        value=""
      />
    </div>
    <div
      class="flex flex-col mb-4"
    >
      <label
        class="uppercase font-bold text-gray-900 mb-2"
        for="name"
      >
        Display Name
      </label>
      <input
        class="border py-2 px-3 text-grey-darkest text-center false"
        id="name"
        placeholder="Your Name"
        type="text"
        value=""
      />
    </div>
    <button
      class="block bg-indigo-800 hover:bg-indigo-700 text-white uppercase text-sm font-semibold mx-auto px-4 py-2 rounded-2xl"
    >
      Join
    </button>
  </div>
</DocumentFragment>
`;
